rfcs/build-system/01_convergence_devstack.md

```diff
@@ -0,0 +1,191 @@
+# WIP | FUI Convergence - devstack/DX
+
+> **tags:** pitch, convergence
+
+> For instructions on shaping a project see here: [Shaping a Project](/kX02SXVbS6KzMOQd56i6Cg)
+>
+> Ported from [hackmd](https://hackmd.io/MiS9ISfjRXqmd6XmxpyMaA)
+
+### Problem
+
+Based on [high level repo issues notes](https://hackmd.io/0TIUjU8_T1Ga2Wjez4IiSQ), we need to address most crucial ones sooner than later to enable good enough DX for both inside/outside contributors.
+
+> Good enough DX:
+>
+> - rapid feedback loop
+> - one common way how to do/run things
+> - docs (reflected within tooling -> `yarn test --help` should just work)
+
+### Appetite
+
+As an initial step lets focus on resolve following (smallest possible area):
+
+1. need to build before run/test/demo
+1. unified TSConfigs
+1. unified way how to transpile javascript
+1. solution for convergence/whole repo
+
+### Solution
+
+> show don't tell / TL;DR
+> ![](https://i.imgur.com/bDwWu9v.png)
+
+#### 1. need to build before run/test/lint/demo
+
+Current state:
+
+- N\* uses wepback aliases with lerna-alias
+- OUIF uses yarn workspaces heavily (symlinks of build assets), thus after every modification outside the package that is being worked on, you need to manually rebuild to get latest changes.
```

Comment:
@ecraig12345 
This isn't correct--the demo experience uses webpack aliases (builds off lerna-alias to also handle deep imports), so you don't have to manually rebuild

@Hotell 
I thought lerna is used only for N*. as you already mentioned lerna is not very well maintained, so I'd recommend to not use it for such things - thus going with path aliases.

Regarding deep imports - correct me if I'm wrong, my knowledge is that we dont wanna support any kind of private apis (deep imports) in convergence.

@ecraig12345 
I'm not sure what you mean "lerna is used only for n*"--I thought it was used together with yarn workspaces for linking in all the packages in our repo. We stopped using it for build orchestration though (in favor of lage).

The definition of "deep imports" gets interesting. In v7/8 we officially support both "@fluentui/pkg" and "@fluentui/pkg/lib/Foo" (plus "@fluentui/pkg/lib/compat/Foo" in a couple special cases) but anything deeper is not supported. (Some of our consumers still use deeper imports because...horrible AMD tooling...but I think we can crack down on this in converged pkgs.)

What I'm not clear on for converged packages is whether we intend to allow imports like "@fluentui/pkg/lib/Foo" internally (or "@fluentui/pkg/src/Foo" and fix the path in the output) since that does require extra configuration. The main benefit of the imports like this is reducing parse time across all tools, which can be a significant issue.

Reply:
Regarding file path imports like `@fluentui/pkg/lib/Foo`, this is an anti-pattern, but has reasons:

1. The top level export doesn't export the thing you need (like a test utility, which like you said, should probably be packaged separately but often customers will work around things if they can, rather than fix things.)
2. The thing customer needs doesn't deserve its own package (questionable)
3. The package you're importing is huge and you're trying to improve build perf as mentioned (primarily cold boot of inner loop, but also CI workflows)
4. You're trying to avoid side effects you absolutely don't need (side effects are the devil)
5. You use AMD and can only resolve file path imports. (still? Surprisingly, yes there are amd websites. Azure portal, OneDrive, and SharePoint are still on require.js)

These are just the reasons; not whether they're right or wrong. File path imports are an anti-pattern, like "using reflection to access private variables" which weren't intended to be accessed. 

I'd love to see it be impossible for customers to do by publishing a rollup, rather than individual JS files.

Reply:
those are great insights @dzearing  thx a bunch!

> The thing customer needs doesn't deserve its own package (questionable)

Right. I think we should be the ones to communicate things explicitly in convergence. What is common in open source ecosystems is to have 2-3 types of packages (contrived example):
- `@fluentui/button` , implementation
- `@fluentui/button-utils` utils when there is some extensive api surface that might be used for instance to build the "implementation" on your own
- `@fluentui/button-testing`, various helpers/tools for testing purposes 

> I'd love to see it be impossible for customers to do by publishing a rollup, rather than individual JS files.

💯. Also this would enable us to ship radically smaller "baggage" to npm registry == faster processing(parse/bundle) within consumers dev stack as well (your no 3. bullet point)

Reply:
Clarifying a couple earlier points about allowed import paths. Today this is what we allow (according to our documentation and partner communication; we don't currently have say an eslint plugin to help enforce):
- `@fluentui/react`
- `@fluentui/react/lib/Foo` where `Foo` is a particular component's root export
- `@fluentui/react/lib/compat`
- `@fluentui/react/lib/compat/Foo` 

In my experience the main (semi) valid justifications for this are @dzearing's points 3-5. The AMD issue being the hardest to work around in other ways. (1) has been allowed in the past but we've been trying to fix these type of issues going forward by properly exporting things people use. 

> > internally (or "@fluentui/pkg/src/Foo" and fix the path in the output) since that does require extra configuration.
> 
> fix the path in the output? can you elaborate ?
Deep imports into other packages within the repo from `src` rather than `lib` have been discussed as a potential way to make intellisense work without prebuild or as much path mapping. Since those paths are not valid for consumers, we'd need a post-build step to replace `src` with `lib`.

> > The main benefit of the imports like this is reducing parse time across all tools, which can be a significant issue.
> 
> That's a hack rather than solution from my experience.
So what is the solution? Is it reasonable to expect all consuming repos (often rather large and burdened by legacy code) to adopt that solution?

> * If you need access non public implementation (deep imports) it should be rather packaged separately, thus having smaller packages with proper API surface.
Agreed, we're working on improving this.

> * Btw do you have any specific numbers regarding "parse time" ? Usually if that's really a big per issue it happens mostly on cold start than current tooling provides good enough caching.
@JasonGore and @kenotron encountered major issues with this in OOUI. One of them might have numbers.

Reply:
Except for the AMD issue, which might require some separate consideration. Shouldn't, for converged components, we try to go forward and disallow/not support file imports? There are good reasons we supported file path imports it in the past, but we seem to agree it causes problems.

It seems like, with smaller more independent packages going forward, it would be easy to address cases 1-4 above with smaller and separate packages for converged components and utilities. 

We do need to consider the AMD case, but perhaps that can be an exception we support for a few partners, rather than affect our main approach.